src/JitInterface/src/SystemVStructClassificator.cs

```diff
@@ -0,0 +1,648 @@
+        /// Returns 'true' if the struct is passed in registers, 'false' otherwise.
+        /// </summary>
+        private static bool ClassifyEightBytes(TypeDesc typeDesc, 
+                                               ref SystemVStructRegisterPassingHelper helper,
+                                               int startOffsetOfStruct)
+        {
+            FieldDesc firstField = null;
+            int numIntroducedFields = 0;
+            foreach (FieldDesc field in typeDesc.GetFields())
+            {
+                if (!field.IsLiteral && !field.IsStatic)
+                {
+                    if (firstField == null)
+                    {
+                        firstField = field;
+                    }
+                    numIntroducedFields++;
+                }
+            }
+
+            if (numIntroducedFields == 0)
+            {
+                return false;
+            }
+
+            // The SIMD Intrinsic types are meant to be handled specially and should not be passed as struct registers
+            if (typeDesc.IsIntrinsic)
+            {
+                InstantiatedType instantiatedType = typeDesc as InstantiatedType;
+                if (instantiatedType != null)
+                {
+                    string typeName = instantiatedType.Name;
+                    string namespaceName = instantiatedType.Namespace;
+
+                    // TODO: is this what we get for the typeName?
+                    if (typeName == "Vector256`1" || typeName == "Vector128`1" || typeName == "Vector64`1")
+                    {
+                        Debug.Assert(namespaceName == "System.Runtime.Intrinsics");
+                        return false;
+                    }
+
+                    if ((typeName ==  "Vector`1") && (namespaceName == "System.Numerics"))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            MetadataType mdType = typeDesc as MetadataType;
+            Debug.Assert(mdType != null);
+
+            TypeDesc firstFieldElementType = firstField.FieldType;
+            int firstFieldSize = firstFieldElementType.GetElementSize().AsInt;
+
+            bool isFixedBuffer = mdType.HasCustomAttribute("System.Runtime.CompilerServices", "FixedBufferAttribute");
+            if (isFixedBuffer)
+            {
+                Debug.Assert(mdType.IsExplicitLayout);
+                numIntroducedFields = typeDesc.GetElementSize().AsInt / firstFieldSize;
+            }
+
+            IEnumerator<FieldDesc> fieldEnumerator = typeDesc.GetFields().GetEnumerator();
+
+            bool hasField = fieldEnumerator.MoveNext();
+            // We've already verified that the type has some fields at the beginning of this function
+            Debug.Assert(hasField);
+
+            FieldAndOffset[] fieldsAndOffsets = mdType.GetClassLayout().Offsets;
+
+            for (int fieldIndex = 0; fieldIndex < numIntroducedFields; fieldIndex++)
```

Comment:
I think the `fieldsAndOffsets` and `fieldEnumerator` (one used for explicit layout, the other used for the rest) is making this too complicated.

Why not:

```csharp
int fieldIndex = 0;
foreach (FieldDesc field in typeDesc.GetFields())
{
    if (field.IsStatic)
        continue;

    fieldIndex++;

    // ...
} 
```

It should come out to fewer lines and is easier to follow (no longer need to special case explicit layout and the "Ignore static and literal fields" disappears).

Reply:
The reason is that for fixed buffers, there is just one field in the typeDesc.GetFields(), but I need to iterate over all the real fields (which for the fixed buffer is `numIntroducedFields` times the same single field).
As for not handling explicit layout - does the fieldsAndOffsets[].Offset match the field.Offset? If it does, then I don't understand why the fieldsAndOffsets entry stores both the Field and Offset and not just the Field.

Reply:
> need to iterate over all the real fields (which for the fixed buffer is numIntroducedFields times the same single field)

I missed the repeated field for fixed buffers. How about

```csharp
int fieldIndex = 0;
foreach (FieldDesc field in isFixedBuffer ? Enumerable.Repeat(firstField, numIntroducedFields) : typeDesc.GetFields())
{
    if (field.IsStatic)
        continue;

    fieldIndex++;

    // ...
} 
```

> does the fieldsAndOffsets[].Offset match the field.Offset? If it does, then I don't understand why the fieldsAndOffsets entry stores both the Field and Offset and not just the Field.

`GetClassLayout` reads what's in the IL metadata. For valuetypes, it will match what you get on `FieldDesc.Offset`. For reference types, it won't match because offsets in metadata always start at zero, but offsets reported by FieldDesc take size of the parent class and padding into account.

`GetClassLayout` is an API that is mostly used when computing the actual field layout. It felt odd to see it here, but it wasn't incorrect because this code deals with valuetypes only.